### PR TITLE
Fix bugs in SearchInput and TextField related to the `label` property

### PR DESCRIPTION
**Pull request by kosmotema**
Starting with MUI 5.4.0, they changed the way of displaying the label for TextField. Now the label will not be rendered only if `null`, `undefined` or `""` value is provided, but react-admin passes `false` as the label (so it will be rendered). This often affects the SearchInput component, which doesn't (and shouldn't) have label: placeholder is shown only when SearchInput is active (because the label replaces the placeholder when the input is not active).

Also, I fixed a bug, where when providing an explicit `undefined` value to the SearchInput's `label` property, the default label (based on `source`) was shown:
```jsx
<SearchInput source="q" label={undefined} />
```

### CODE REVIEW

**Review by WiXSL — APPROVED** [review]

**Comment by WiXSL** [comment]
Thank you!

File: packages/ra-ui-materialui/src/input/SearchInput.tsx [hunk 1]
```diff
@@ -8,17 +8,18 @@ import { CommonInputProps } from './CommonInputProps';
 import { TextInput, TextInputProps } from './TextInput';
 
 export const SearchInput = (props: SearchInputProps) => {
+    const { label, ...rest } = props;
```
Comment (fzaninotto) [inline packages/ra-ui-materialui/src/input/SearchInput.tsx:11]: I don't understand why you made this change

File: packages/ra-ui-materialui/src/input/SearchInput.tsx [hunk 2]
```diff
@@ -8,17 +8,18 @@ import { CommonInputProps } from './CommonInputProps';
...
-    if (props.label) {
+    if (label) {
         throw new Error(
             "<SearchInput> isn't designed to be used with a label prop. Use <TextInput> if you need a label."
         );
     }
 
     return (
         <StyledTextInput
-            hiddenLabel
```
Comment (fzaninotto) [inline packages/ra-ui-materialui/src/input/SearchInput.tsx:21]: why did you remove that?

File: packages/ra-ui-materialui/src/input/TextInput.spec.tsx [hunk 3]
```diff
@@ -166,4 +166,36 @@ describe('<TextInput />', () => {
             expect(input.value).toEqual('bar');
         });
     });
+
+    describe('label', () => {
+        it('should not render label when `label` prop is `false`', () => {
```
Comment (fzaninotto) [inline packages/ra-ui-materialui/src/input/TextInput.spec.tsx:171]: You're missing a test showing that a proper label prop renders a label.

File: packages/ra-ui-materialui/src/input/SearchInput.spec.tsx [hunk 4]
```diff
@@ -0,0 +1,30 @@
+import * as React from 'react';
+import { render } from '@testing-library/react';
+import { testDataProvider } from 'ra-core';
+
+import { AdminContext } from '../AdminContext';
+import { SearchInput } from '.';
+import { FilterForm } from '../list';
+
+describe('<SearchInput />', () => {
+    it('should not render label if passed explicit `undefined` value', async () => {
```
Comment (fzaninotto) [inline packages/ra-ui-materialui/src/input/SearchInput.spec.tsx:10]: I don't understand this test.

**Review by fzaninotto — CHANGES_REQUESTED** [review]
Could you show before / after screenshots?

Also, I understand that MUI changed, but I want to be sure that react-admin works with both the old and the new MUI versions. Are you sure this change is backwards compatible?

File: packages/ra-ui-materialui/src/input/SearchInput.tsx (hunk 1 above)
Comment (kosmotema) [reply packages/ra-ui-materialui/src/input/SearchInput.tsx:11]: To handle the case when specifying `undefined` for a SearchInput's label explicitly (like `<SearchInput source="q" alwaysOn label={undefined} />`) will result in the default label is being displayed.
[Reproduction on codesandbox](https://codesandbox.io/s/react-admin-explicit-undefined-label-miiic7?file=/src/posts/PostList.tsx)

File: packages/ra-ui-materialui/src/input/SearchInput.tsx (hunk 2 above)
Comment (kosmotema) [reply packages/ra-ui-materialui/src/input/SearchInput.tsx:21]: My bad, added it back

File: packages/ra-ui-materialui/src/input/TextInput.spec.tsx (hunk 3 above)
Comment (kosmotema) [reply packages/ra-ui-materialui/src/input/TextInput.spec.tsx:201]: I've added two more tests: when `label` not specified and when `label` is a non-empty string

File: packages/ra-ui-materialui/src/input/SearchInput.spec.tsx (hunk 4 above)
Comment (kosmotema) [reply packages/ra-ui-materialui/src/input/SearchInput.spec.tsx:10]: Added to cover the case of passing explicit `undefined` for the SearchInput's label property (more info [here](https://github.com/marmelab/react-admin/pull/8439#discussion_r1033939044))

**Comment by kosmotema** [comment]
> Could you show before / after screenshots?

I made a reproduction of the bug:
1. [With MUI 5.0](https://codesandbox.io/s/react-admin-mui-5-0-8yp49i)
2. [With MUI 5.4](https://codesandbox.io/s/react-admin-mui-5-4-37ivhr)
3. [With MUI 5.0 and patched react-admin](https://codesandbox.io/s/react-admin-patched-mui-5-0-7hgvc3)
4. [With MUI 5.4 and patched react-admin](https://codesandbox.io/s/react-admin-patched-mui-5-4-6p0ef9)

And with screenshots:
The placeholder is not displayed when using MUI > 5.4 and latest react-admin (because MUI [due to changed logic of displaying a label] renders the `label` element with empty text):

mui > 5.4, stable react-admin, placeholder is hidden

But the placeholder is visible after clicking on the SearchInput:

placeholder is visible

With the proposed changes, the placeholder now displayed correctly (both until the SearchInput is focused, and after that):

placholder is visible, SearchInput not focused

placeholder is also visiblde, SearchInput as focused

File: packages/ra-ui-materialui/src/input/SearchInput.tsx (hunk 1 above)
Comment (kosmotema) [reply packages/ra-ui-materialui/src/input/SearchInput.tsx:11]: Also noticed, that it's possible to use SearchInput with `0` label like:
```js
<SearchInput source="q" alwaysOn label={0} />
```

I think, it would be better to revert all changes in this PR, that related to the SearchInput's label overwriting logic and open a new issue to discuss, what and how to do about this case (IMHO it's better to prevent users from providing a `label` at all instead of throwing an error).

**Review by slax57 — APPROVED** [review]

**Comment by slax57** [comment]
Tested the fix in `TextInput` successfully, and the change in `SearchInput` is harmless, so I'll merge this PR as is.

Thanks for this contribution and well done for the analysis!
